Linker fails with "cannot find -lpython3.9" error during cross-compile

I'm using buildroot to cross-compile an image for the Beaglebone Black (Arm).  I've gotten cyclonedds to install in BR; now I'm trying to get the cyclonedds-python binding installed.  BR is able to get through compiling, but when linking, it fails saying it 'cannot find -lpython3.9', thusly:

[100%] Linking C shared module _clayer.cpython-39-arm-linux-gnueabihf.so
/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/gcc/arm-buildroot-linux-gnueabihf/9.4.0/../../../../arm-buildroot-linux-gnueabihf/bin/ld: cannot find -lpython3.9
collect2: error: ld returned 1 exit status
make[4]: *** [CMakeFiles/_clayer.dir/build.make:119: _clayer.cpython-39-arm-linux-gnueabihf.so] Error 1
make[3]: *** [CMakeFiles/Makefile2:95: CMakeFiles/_clayer.dir/all] Error 2
make[2]: *** [Makefile:149: all] Error 2
Traceback (most recent call last):
  File "/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/python3.9/site-packages/skbuild/setuptools_wrap.py", line 593, in setup
    cmkr.make(make_args, env=env)
  File "/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/python3.9/site-packages/skbuild/cmaker.py", line 504, in make
    raise SKBuildError(

An error occurred while building with CMake.
  Command:
    cmake --build . --target install --config Release --
  Source directory:
    /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996
  Working directory:
    /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/_skbuild/linux-x86_64-3.9/cmake-build
Please see CMake's output for more information.
make[1]: *** [package/pkg-generic.mk:251: /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/.stamp_built] Error 1
make: *** [Makefile:84: _all] Error 2

For the record, python 3.9 is the version being used in my BR image.

Any suggestions on what is causing this?

So far I've not been able to make it work, i.e., get cyclonedds-python to link under buildroot.  Adding the python-dev package, as one would do on a running linux distro, doesn't really make sense in buildroot because BR creates all its packages from source code.  So, there are no *-dev packages in BR.

Packages in BR are generally built following the _configure, make, make install_ path.  It also has frameworks for packages that are built & installed via CMake, or in the case of python packages, via setuptools or distutils although even with these it seems to ultimately run CMake which means it ultimately is running make.  To create a new package, a <package>.mk file is created containing some variables and the compilation method (setuptools in the case of cyclonedds-python).  (FYI, under BR, cyclonedds-python is called python-cyclonedds.)  For example, to get python-cyclonedds to compile, I had to add the variable PYTHON_CYCLONEDDS_ENV = CMAKE_PREFIX_PATH=$(STAGING_DIR)/ to the python-cyclonedds.mk file.  To get the linking to work, I tried adding CMAKE_LIBRARY_PATH=$(STAGING_DIR)/  and then CMAKE_SYSTEM_LIBRARY_PATH=$(STAGING_DIR)/ to PYTHON_CYCLONEDDS_ENV, but neither made any difference.

I've also tried adding LIBRARY_PATH to PTYHON_CYCLONEDDS_ENV, but I'm unsure of what the value should be.  The python3.9 directory is in $(STAGING_DIR)/usr/lib/, so I've tried setting it to just $(STAGING_DIR)/, and also to $(STAGING_DIR)/usr/lib, but in both cases, I still get the same 'cannot find -lpython3.9' error.

Looking through menuconfig, I did notice that the "python3 module format to install" is set to ".pyc compiled sources only"; should that be changed to ".py sources and .pyc compiled"?  I'll try that and report back.

Sadly, switching to '.py sources and .pyc compiled" for python modules made no difference and I get the same "cannot find -lpython3.9" error during linking.  I'm not surprised because I don't think the issue is that the modules/libraries are not there, it's that ld can't find them.  I don't know much about ld, but where is the path defined that tells it where to look?  The man page for ld says -l  is for specifying an archive or object file; but $(STAGING_DIR)/usr/lib/python3.9 is a directory, and it contains a bunch of python modules in .py and .pyc format, not object files.  Am I pointing ld to the wrong directory?

For the record, here is the complete log of the attempt to get cyclonedds-python (aka python-cyclonedds) to build under buildroot (just in case someone spots an earlier incident that would explain the problem):

>>> python-cyclonedds 1e1af632cf04403b7f372a55a74b2b76da1ba996 Extracting
gzip -d -c /home/petre/buildroot-2021.02.8.beaglebone/dl/python-cyclonedds/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996-br1.tar.gz | tar --strip-components=1 -C /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996   -xf -
>>> python-cyclonedds 1e1af632cf04403b7f372a55a74b2b76da1ba996 Patching
>>> python-cyclonedds 1e1af632cf04403b7f372a55a74b2b76da1ba996 Configuring
>>> python-cyclonedds 1e1af632cf04403b7f372a55a74b2b76da1ba996 Building
(cd /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996//; _PYTHON_SYSCONFIGDATA_NAME="`{ [ -e /home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/lib/python3.9//_sysconfigdata__linux_*.py ] && basename /home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/lib/python3.9//_sysconfigdata__linux_*.py .py; } || true`" PATH="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin:/home/petre/buildroot-2021.02.8.beaglebone/output/host/sbin:/opt/cmake-3.18.6/bin:/home/petre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/cyclonedds/bin:/opt/cyclonedds/bin" PATH="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin:/home/petre/buildroot-2021.02.8.beaglebone/output/host/sbin:/opt/cmake-3.18.6/bin:/home/petre/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/opt/cyclonedds/bin:/opt/cyclonedds/bin" AR="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-ar" AS="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-as" LD="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-ld" NM="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-nm" CC="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gcc" GCC="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gcc" CPP="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-cpp" CXX="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-g++" FC="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gfortran" F77="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gfortran" RANLIB="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-ranlib" READELF="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-readelf" STRIP="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-strip" OBJCOPY="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-objcopy" OBJDUMP="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-objdump" AR_FOR_BUILD="/usr/bin/ar" AS_FOR_BUILD="/usr/bin/as" CC_FOR_BUILD="/usr/bin/gcc" GCC_FOR_BUILD="/usr/bin/gcc" CXX_FOR_BUILD="/usr/bin/g++" LD_FOR_BUILD="/usr/bin/ld" CPPFLAGS_FOR_BUILD="-I/home/petre/buildroot-2021.02.8.beaglebone/output/host/include" CFLAGS_FOR_BUILD="-O2 -I/home/petre/buildroot-2021.02.8.beaglebone/output/host/include" CXXFLAGS_FOR_BUILD="-O2 -I/home/petre/buildroot-2021.02.8.beaglebone/output/host/include" LDFLAGS_FOR_BUILD="-L/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib -Wl,-rpath,/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib" FCFLAGS_FOR_BUILD="" DEFAULT_ASSEMBLER="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-as" DEFAULT_LINKER="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-ld" CPPFLAGS="-D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64" CFLAGS="-D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64  -Os  " CXXFLAGS="-D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64  -Os  " LDFLAGS="" FCFLAGS=" -Os " FFLAGS=" -Os " PKG_CONFIG="/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/pkg-config" STAGING_DIR="/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot" INTLTOOL_PERL=/usr/bin/perl PYTHONPATH="/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/lib/python3.9/" PYTHONNOUSERSITE=1 _python_sysroot=/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot _python_prefix=/usr _python_exec_prefix=/usr CMAKE_PREFIX_PATH=/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/ CMAKE_LIBRARY_PATH=/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/ LIBRARY_PATH=/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/ CMAKE_SYSTEM_LIBRARY_PATH=/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/ /home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/python setup.py build  )


--------------------------------------------------------------------------------
-- Trying "Ninja" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
-- Configuring incomplete, errors occurred!
See also "/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/_cmake_test_compile/build/CMakeFiles/CMakeOutput.log".
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Ninja" generator - failure
--------------------------------------------------------------------------------



--------------------------------------------------------------------------------
-- Trying "Unix Makefiles" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- The C compiler identification is GNU 9.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is GNU 9.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/_cmake_test_compile/build
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Unix Makefiles" generator - success
--------------------------------------------------------------------------------

Configuring Project
  Working directory:
    /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/_skbuild/linux-x86_64-3.9/cmake-build
  Command:
    cmake /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996 -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX:PATH=/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/_skbuild/linux-x86_64-3.9/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/python -DPYTHON_VERSION_STRING:STRING=3.9.7 -DPYTHON_INCLUDE_DIR:PATH=/home/petre/buildroot-2021.02.8.beaglebone/output/host/include/python3.9 -DPYTHON_LIBRARY:FILEPATH=libpython3.9.a -DSKBUILD:INTERNAL=TRUE -DCMAKE_MODULE_PATH:PATH=/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/python3.9/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release

-- The C compiler identification is GNU 9.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PythonInterp: /home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/python (found version "3.9.7") 
-- Found PythonLibs: libpython3.9.a (found version "3.9.7") 
-- Performing Test Weak Link MODULE -> SHARED (gnu_ld_ignore) - Failed
-- Performing Test Weak Link MODULE -> SHARED (osx_dynamic_lookup) - Failed
-- Performing Test Weak Link MODULE -> SHARED (no_flag) - Failed
_modinit_prefix:PyInit_
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    SKBUILD


-- Build files have been written to: /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/_skbuild/linux-x86_64-3.9/cmake-build
Scanning dependencies of target _clayer
[ 33%] Building C object CMakeFiles/_clayer.dir/clayer/cdrkeyvm.c.o
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c: In function ‘cdr_key_vm_run’:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c:384:61: warning: left shift count >= width of type [-Wshift-count-overflow]
  384 |                     ((size_t)*(cdr_sample + sample_pos + 4) << 32) | ((size_t)*(cdr_sample + sample_pos + 5) << 40) |
      |                                                             ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c:384:110: warning: left shift count >= width of type [-Wshift-count-overflow]
  384 |                     ((size_t)*(cdr_sample + sample_pos + 4) << 32) | ((size_t)*(cdr_sample + sample_pos + 5) << 40) |
      |                                                                                                              ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c:385:61: warning: left shift count >= width of type [-Wshift-count-overflow]
  385 |                     ((size_t)*(cdr_sample + sample_pos + 6) << 48) | ((size_t)*(cdr_sample + sample_pos + 7) << 56) :
      |                                                             ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c:385:110: warning: left shift count >= width of type [-Wshift-count-overflow]
  385 |                     ((size_t)*(cdr_sample + sample_pos + 6) << 48) | ((size_t)*(cdr_sample + sample_pos + 7) << 56) :
      |                                                                                                              ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c:388:61: warning: left shift count >= width of type [-Wshift-count-overflow]
  388 |                     ((size_t)*(cdr_sample + sample_pos + 3) << 32) | ((size_t)*(cdr_sample + sample_pos + 2) << 40) |
      |                                                             ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c:388:110: warning: left shift count >= width of type [-Wshift-count-overflow]
  388 |                     ((size_t)*(cdr_sample + sample_pos + 3) << 32) | ((size_t)*(cdr_sample + sample_pos + 2) << 40) |
      |                                                                                                              ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c:389:61: warning: left shift count >= width of type [-Wshift-count-overflow]
  389 |                     ((size_t)*(cdr_sample + sample_pos + 1) << 48) | ((size_t)*(cdr_sample + sample_pos) << 56);
      |                                                             ^~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/cdrkeyvm.c:389:106: warning: left shift count >= width of type [-Wshift-count-overflow]
  389 |                     ((size_t)*(cdr_sample + sample_pos + 1) << 48) | ((size_t)*(cdr_sample + sample_pos) << 56);
      |                                                                                                          ^~
[ 66%] Building C object CMakeFiles/_clayer.dir/clayer/pysertype.c.o
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c:651:1: warning: missing initializer for field ‘derive_sertype’ of ‘const struct ddsi_sertype_ops’ [-Wmissing-field-initializers]
  651 | };
      | ^
In file included from /home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/include/dds/ddsi/ddsi_serdata.h:18,
                 from /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c:25:
/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/include/dds/ddsi/ddsi_sertype.h:169:25: note: ‘derive_sertype’ declared here
  169 |   ddsi_sertype_derive_t derive_sertype;
      |                         ^~~~~~~~~~~~~~
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c: In function ‘ddspy_read_handle’:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c:1120:30: warning: conversion to ‘Py_ssize_t’ {aka ‘int’} from ‘uint32_t’ {aka ‘unsigned int’} may change the sign of the result [-Wsign-conversion]
 1120 |         PyList_SetItem(list, i, item); // steals ref
      |                              ^
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c: In function ‘ddspy_take_handle’:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c:1166:30: warning: conversion to ‘Py_ssize_t’ {aka ‘int’} from ‘uint32_t’ {aka ‘unsigned int’} may change the sign of the result [-Wsign-conversion]
 1166 |         PyList_SetItem(list, i, item); // steals ref
      |                              ^
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c: In function ‘ddspy_read_participant’:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c:1455:30: warning: conversion to ‘Py_ssize_t’ {aka ‘int’} from ‘uint32_t’ {aka ‘unsigned int’} may change the sign of the result [-Wsign-conversion]
 1455 |         PyList_SetItem(list, i, item); // steals ref
      |                              ^
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c: In function ‘ddspy_take_participant’:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c:1508:30: warning: conversion to ‘Py_ssize_t’ {aka ‘int’} from ‘uint32_t’ {aka ‘unsigned int’} may change the sign of the result [-Wsign-conversion]
 1508 |         PyList_SetItem(list, i, item); // steals ref
      |                              ^
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c: In function ‘ddspy_read_endpoint’:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c:1570:30: warning: conversion to ‘Py_ssize_t’ {aka ‘int’} from ‘uint32_t’ {aka ‘unsigned int’} may change the sign of the result [-Wsign-conversion]
 1570 |         PyList_SetItem(list, i, item); // steals ref
      |                              ^
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c: In function ‘ddspy_take_endpoint’:
/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/clayer/pysertype.c:1658:30: warning: conversion to ‘Py_ssize_t’ {aka ‘int’} from ‘uint32_t’ {aka ‘unsigned int’} may change the sign of the result [-Wsign-conversion]
 1658 |         PyList_SetItem(list, i, item); // steals ref
      |                              ^
[100%] Linking C shared module _clayer.cpython-39-arm-linux-gnueabihf.so
/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/gcc/arm-buildroot-linux-gnueabihf/9.4.0/../../../../arm-buildroot-linux-gnueabihf/bin/ld: cannot find -lpython3.9
collect2: error: ld returned 1 exit status
make[4]: *** [CMakeFiles/_clayer.dir/build.make:119: _clayer.cpython-39-arm-linux-gnueabihf.so] Error 1
make[3]: *** [CMakeFiles/Makefile2:95: CMakeFiles/_clayer.dir/all] Error 2
make[2]: *** [Makefile:149: all] Error 2
Traceback (most recent call last):
  File "/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/python3.9/site-packages/skbuild/setuptools_wrap.py", line 593, in setup
    cmkr.make(make_args, env=env)
  File "/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/python3.9/site-packages/skbuild/cmaker.py", line 504, in make
    raise SKBuildError(

An error occurred while building with CMake.
  Command:
    cmake --build . --target install --config Release --
  Source directory:
    /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996
  Working directory:
    /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/_skbuild/linux-x86_64-3.9/cmake-build
Please see CMake's output for more information.
make[1]: *** [package/pkg-generic.mk:251: /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-1e1af632cf04403b7f372a55a74b2b76da1ba996/.stamp_built] Error 1
make: *** [Makefile:84: _all] Error 2

Looking at the CMakeLists.txt file, it appears the failure happens at line 32, `target_link_libraries(_clayer CycloneDDS::ddsc)`.  Where is the "python3.9" argument actually passed to ld?

Hi @pscheie, what stands out for me is this:
> /home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/gcc/arm-buildroot-linux-gnueabihf/9.4.0/../../../../arm-buildroot-linux-gnueabihf/bin/ld: cannot find -lpython3.9

That means the linker can't find a `libpython3.9.so` or `libpython3.9.a` in the linker path, which is really the python interpreter proper (in this case, for the target platform). A module needs to link to it because it needs to call various python routines to read/create/modify the python objects that it works with.

The linker library path is set the `-L` option. If you do `make -DVERBOSE=1` you'll should see all the commands executed by `make` during the build, and so it would be interesting to see what the library path is for that command. The other two things that are interesting are:
* Where is the (target's) `libpython3.9.a` (I think it is looking for a `.a`)? `find / -name libpython3.9.a` should find it if it is somewhere in the filesystem, but it will also find the host's version of it. The path will probably tell you which is the right one, else the `file` command can help you figure it out. (If you have a good guess where it might be, try that first instead of the starting in `/` :))
* Assuming it is present, where is the linker path set? It may be that you can fix it with some cmake magic (hopefully some kind soul can help you there), else, as thing of last resort, there is the possibility of editing the *output* of CMake. I've done it in the past, IIRC the key bit is in a file named `link.txt`, but it is tricky, unpleasant, and undone by running cmake again ... so far from ideal.

A workaround for fixing the path passed to the linker is to put a soft-link in a place where it does look. Also, it looks like there is a define for what it should link to (`-DPYTHON_LIBRARY:FILEPATH=libpython3.9.a`) so overriding that might help. And finally it includes the python headers from `/home/petre/buildroot-2021.02.8.beaglebone/output/host/include/python3.9` which may be right but seems weird to me for a cross-build ...

I'm sorry I can't really help you ...

Hi @eboasson,
Thanks for the response.  I misspoke about how buildroot tries to build python packages; it calls setup.py, although in the case of cyclonedds-python, that ends up calling cmake, right?

Does cyclonedds-python define a linker library path, or just assume the default path defined within the Python environment?

What is the difference between running 'pip install' vs 'python setup.py install' ?  (I assume BR does the latter.)

BTW, I tried adding -DVERBOSE=1 to my call to make but it doesn't recognize that option.  -d gives me debug output but there's no additional info around the 'ld: cannot find -lpython3.9' error.

Hi @thijsmie,
You can download the file from http://peterscheie.com/files/buildroot-2021.02.8.beaglebone.thijsmie.tgz
Untar the tar file and cd into buildroot-2021.02.8.beaglebone/.  Here are the steps I use to build right now (it's rather ugly):

1. As I'm sure you know, you need to have cyclonedds installed on the host system somewhere.  I have it in /opt/cyclonedds, and the package definition for python-cyclonedds references that so at least have /opt/cyclonedds as a symlink to your cyclonedds or modify package/python-cyclonedds/python-cyclonedds.mk to point to your cyclonedds directory.
2. Run 'make all'.
3. You may see a couple of warnings about cmake 3.18 being an illegal version number; ignore them.
3. After about 20 minutes the build will fail while trying to build cyclonedds because there should be a LICENSE file in output/build/cyclonedds-ec866a11f66bf7cfac02b024823d92368d4049c8/ and for some reason it gets renamed to license.txt as described in [https://github.com/eclipse-cyclonedds/cyclonedds/issues/1142](url).  I just rename the file back to LICENSE, and then start 'make all' again.
4. After several more minutes, the build will fail when trying to build python-cyclonedds (the buildroot name I gave to cyclonedds-python, as all python packages in BR start their names with 'python-*') because of the 'ld cannot find -lpython3.9' error.  Edit output/build/python-cyclonedds-6de6977210b392ae34c2f732006952ee4608ca13/_skbuild/linux-x86_64-3.9/cmake-build/CMakeCache.txt and at line 207 change 
`PYTHON_LIBRARY:FILEPATH=libpython3.9.a` 
to 
`PYTHON_LIBRARY:FILEPATH=<path to buildroot>/buildroot-2021.02.8.beaglebone/output/target/usr/lib/python3.9/config-3.9-arm-linux-gnueabihf/libpython3.9.a`.  Save & exit the file.  (Alternatively, there's also a patch file, 0001-fix-path-to-libpython3.9.a-in-CMakeCache.txt.patch and a shell script, fix-libpython.sh, in the buildroot directory that I use to change the path; but the path specified in the patch will probably be different than where you have buildroot setup, i.e., it won't be in /home/petre/.)
Anyway, after changing the path to libpython3.9.a, start 'make all' again.

Make will run for several more minutes, but then it should finish successfully.  Look in output/images/ for the resulting images.  To really test the image(s) you'll need a BeagleBone Black (last time I checked, Qemu didn't support emulating a BBB).  In the quite-likely event there are troubles getting the build to run to completion, let me know.

Hi @thijsmie,
I changed the python-cyclonedds.mk to pull from your branch and pointed CYCLONEDDS_HOME to $(TARGET_DIR)/ as you suggested.  Now the linker is failing to find cycloneddsidl:

/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Os -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Os -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -fPIC -I/home/petre/buildroot-2021.02.8.beaglebone/output/target/include -I/home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-no-scikit-build/clayer -I/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/include/python3.9 -c clayer/typeser.c -o build/temp.linux-x86_64-3.9/clayer/typeser.o
/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gcc -shared -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Os -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 build/temp.linux-x86_64-3.9/clayer/cdrkeyvm.o build/temp.linux-x86_64-3.9/clayer/pysertype.o build/temp.linux-x86_64-3.9/clayer/typeser.o -L/home/petre/buildroot-2021.02.8.beaglebone/output/target/lib -L/home/petre/buildroot-2021.02.8.beaglebone/output/target/lib64 -L/home/petre/buildroot-2021.02.8.beaglebone/output/target/bin -L/home/petre/buildroot-2021.02.8.beaglebone/output/host/arm-buildroot-linux-gnueabihf/sysroot/usr/lib -lcycloneddsidl -lddsc -o build/lib.linux-x86_64-3.9/cyclonedds/_clayer.cpython-39-arm-linux-gnueabihf.so
**/home/petre/buildroot-2021.02.8.beaglebone/output/host/lib/gcc/arm-buildroot-linux-gnueabihf/9.4.0/../../../../arm-buildroot-linux-gnueabihf/bin/ld: cannot find -lcycloneddsidl**
collect2: error: ld returned 1 exit status
error: command '/home/petre/buildroot-2021.02.8.beaglebone/output/host/bin/arm-buildroot-linux-gnueabihf-gcc' failed with exit code 1
make[1]: *** [package/pkg-generic.mk:251: /home/petre/buildroot-2021.02.8.beaglebone/output/build/python-cyclonedds-no-scikit-build/.stamp_built] Error 1
make: *** [Makefile:84: _all] Error 2

There's no cycloneddsidl* file in the target/ directory, nor even in host system's /opt/cyclonedds/ directory (not that the build would use that, but presumably it was needed when I installed cyclonedds-python on the host system).  Ideas?

--- a/.azure/publish-release.yml
+++ b/.azure/publish-release.yml
@@ -38,7 +38,7 @@ variables:
   CONFIGURE: >-
     mkdir -p build &&
     cd build &&
-    cmake ../main -DCMAKE_INSTALL_PREFIX="$CYCLONEDDS_HOME" -DBUILD_SCHEMA=off -DENABLE_SHM=off -DENABLE_SSL=off -DENABLE_SECURITY=off -DENABLE_TOPIC_DISCOVERY=on -DENABLE_TYPE_DISCOVERY=on
+    cmake ../main -DCMAKE_INSTALL_PREFIX="$CYCLONEDDS_HOME" -DBUILD_SCHEMA=off -DENABLE_SHM=off -DENABLE_SSL=on -DENABLE_SECURITY=on -DENABLE_TOPIC_DISCOVERY=on -DENABLE_TYPE_DISCOVERY=on
   BUILD: >-
     cmake --build . --target install
   CIBW_BEFORE_ALL: >-
@@ -48,16 +48,16 @@ variables:
     auditwheel repair -w {dest_dir} {wheel}
   CIBW_ENVIRONMENT_LINUX: >-
     CYCLONEDDS_HOME=/project/install
-    CMAKE_PREFIX_PATH=/project/install
     LD_LIBRARY_PATH=/project/install/lib:/project/install/lib64
+    STANDALONE_WHEELS=1
   # ---- MACOS ----
   CIBW_REPAIR_WHEEL_COMMAND_MACOS: >-
     DYLD_LIBRARY_PATH=$(Build.Repository.LocalPath)/install/lib delocate-listdeps {wheel} &&
     DYLD_LIBRARY_PATH=$(Build.Repository.LocalPath)/install/lib delocate-wheel --require-archs {delocate_archs} -w {dest_dir} {wheel}
   CIBW_ENVIRONMENT_MACOS: >-
     CYCLONEDDS_HOME=$(Build.Repository.LocalPath)/install
-    CMAKE_PREFIX_PATH=$(Build.Repository.LocalPath)/install
     DYLD_LIBRARY_PATH=$(Build.Repository.LocalPath)/install/lib
+    STANDALONE_WHEELS=1
   # ---- WINDOWS ----
   CIBW_REPAIR_WHEEL_COMMAND_WINDOWS: >-
     pip install delvewheel==0.0.14 &&
@@ -93,9 +93,9 @@ jobs:
          $(CLONE) && $(CONFIGURE) -A Win32 && $(BUILD)
   steps:
   - pwsh: |
+      Write-Host "###vso[task.setvariable variable=STANDALONE_WHEELS;]1"
       Write-Host "###vso[task.setvariable variable=PATH;]$(Build.Repository.LocalPath)\\install\\bin;${env:PATH}"
       Write-Host "###vso[task.setvariable variable=CYCLONEDDS_HOME;]$(Build.Repository.LocalPath)\install"
-      Write-Host "###vso[task.setvariable variable=CMAKE_PREFIX_PATH;]$(Build.Repository.LocalPath)\install"
       $a = "${env:CIBW_BEFORE_ALL_WINDOWS_PRE}" -replace "\`$CYCLONEDDS_HOME","$(Build.Repository.LocalPath)\install"
       Write-Host "###vso[task.setvariable variable=CIBW_BEFORE_ALL_WINDOWS;]$a"
     condition: eq(variables['Agent.OS'], 'Windows_NT')
@@ -128,4 +128,3 @@ jobs:
       find artifacts/ -name "*.whl" -exec mv {} wheelhouse \;
     displayName: Collect wheels for publication
   - template: /.azure/templates/publish-package.yml
-

--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,6 @@
+# Custom
+cyclonedds/__library__.py
+
 # Byte-compiled / optimized / DLL files
 __pycache__/
 *.py[cod]

--- a/README.md
+++ b/README.md
@@ -41,7 +41,7 @@ $ pip3 install --user .[dev]
 $ pip3 install --user .[docs]
 ```
 
-While the quickest way to get going is the `--user` flag it is not the recommended, we recommend using [a virtual environment][2], [poetry][3], [pipenv][4] or [pyenv][5]. After the installation is complete `import cyclonedds` should now work. The `CYCLONEDDS_HOME` variable is essential for the Python backend to locate the CycloneDDS binaries so this always needs to be set when running Python code with Cyclone DDS. Have a look at the [examples](examples/) to learn about how to use the Python API.
+While the quickest way to get going is the `--user` flag it is not the recommended, we recommend using [a virtual environment][2], [poetry][3], [pipenv][4] or [pyenv][5]. After the installation is complete `import cyclonedds` should now work. Have a look at the [examples](examples/) to learn about how to use the Python API.
 
 You can also run the idl compiler in Python mode if it the Python package is installed. Simply run `idlc -l py file.idl` and a Python module with your types will be generated in the current working directory. If you wish to nest the resulting Python module inside an existing package you can specify the path from the intended root. So if you have a package 'wubble' with a submodule 'fruzzy' and want the generated modules and types under there you can do `idlc -l py -p py-root-prefix=wubble.fruzzy file.idl`.
 

--- a/buildhelp/bdist_wheel.py
+++ b/buildhelp/bdist_wheel.py
@@ -1,0 +1,39 @@
+from wheel.bdist_wheel import bdist_wheel as _bdist_wheel
+from cyclone_search import find_cyclonedds
+from pathlib import Path
+import shutil
+import os
+
+
+class bdist_wheel(_bdist_wheel):
+    def initialize_options(self):
+        self.standalone = os.environ.get("STANDALONE_WHEELS") == "1"
+        super().initialize_options()
+
+    def finalize_options(self):
+        if self.standalone:
+            self.distribution.entry_points["console_scripts"].append("idlc=cyclonedds.tools.wheel_idlc:command")
+        super().finalize_options()
+
+    def run(self):
+        if self.standalone:
+            cyclone = find_cyclonedds()
+
+            newlibdir = Path(self.bdist_dir) / 'cyclonedds' / '.libs'
+
+            os.makedirs(newlibdir, exist_ok=True)
+            with open(Path(self.bdist_dir) / 'cyclonedds' / '__library__.py', "w", encoding='utf-8') as f:
+                f.write("from pathlib import Path\n\n")
+                f.write("in_wheel = True\n")
+                f.write(f"library_path = Path(__file__).parent / '.libs' / '{cyclone.ddsc_library.name}'")
+
+            shutil.copy(cyclone.ddsc_library, newlibdir / cyclone.ddsc_library.name)
+
+            if cyclone.idlc_executable and cyclone.idlc_library:
+                shutil.copy(cyclone.idlc_executable, newlibdir / cyclone.idlc_executable.name)
+                shutil.copy(cyclone.idlc_library, newlibdir / cyclone.idlc_library.name)
+
+            for lib in cyclone.security_libs:
+                shutil.copy(lib, newlibdir / lib.name)
+
+        super().run()

--- a/buildhelp/build_ext.py
+++ b/buildhelp/build_ext.py
@@ -1,0 +1,19 @@
+from setuptools.command.build_ext import build_ext as _build_ext
+from setuptools import Extension
+
+
+class Library(Extension):
+    pass
+
+
+class build_ext(_build_ext):
+    def get_libraries(self, ext):
+        if isinstance(ext, Library):
+            return ext.libraries
+        return super().get_libraries(ext)
+
+    def get_export_symbols(self, ext):
+        if isinstance(ext, Library):
+            return ext.export_symbols
+        return super().get_export_symbols(ext)
+

--- a/buildhelp/cyclone_search.py
+++ b/buildhelp/cyclone_search.py
@@ -1,0 +1,121 @@
+"""
+ * Copyright(c) 2022 ZettaScale Technology and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+"""
+import os
+import platform
+from pathlib import Path
+from typing import List, Optional
+from dataclasses import dataclass
+
+
+@dataclass
+class FoundCycloneResult:
+    home: Path
+    library_path: Path
+    include_path: Path
+    binary_path: Path
+    ddsc_library: Path
+    idlc_executable: Optional[Path]
+    idlc_library: Optional[Path]
+    security_libs: List[Path]
+
+
+def first_or_none(alist):
+    alist = list(alist)
+    if alist:
+        return alist[0]
+
+
+def good_directory(directory: Path):
+    dir = directory.resolve()
+    if not dir.exists():
+        return
+
+    include_path = dir / 'include'
+    bindir = dir / 'bin'
+
+    if not include_path.exists() or not bindir.exists():
+        return
+
+    libdir = dir / 'lib'
+    if not libdir.exists():
+        libdir = dir / 'lib64'
+        if not libdir.exists():
+            return None
+
+    if platform.system() == 'Windows':
+        ddsc_library = bindir / "ddsc.dll"
+    elif platform.system() == 'Darwin':
+        ddsc_library = libdir / "libddsc.dylib"
+    else:
+        ddsc_library = libdir / "libddsc.so"
+
+    if not ddsc_library.exists():
+        return None
+
+    idlc_executable = first_or_none(bindir.glob("idlc*"))
+    idlc_library = first_or_none(libdir.glob('libcycloneddsidl*')) or first_or_none(bindir.glob("cycloneddsidl*"))
+    security_libs = list(libdir.glob("*dds_security_*")) + list(bindir.glob("*dds_security_*"))
+
+    return FoundCycloneResult(
+        home=dir,
+        include_path=include_path,
+        library_path=libdir,
+        binary_path=bindir,
+        ddsc_library=ddsc_library,
+        idlc_executable=idlc_executable,
+        idlc_library=idlc_library,
+        security_libs=security_libs
+    )
+
+
+def search_cyclone_pathlike(pathlike, upone=False):
+    for path in pathlike.split(os.pathsep):
+        if upone:
+            return good_directory(Path(path) / '..')
+        else:
+            return good_directory(Path(path))
+
+
+def find_cyclonedds() -> Optional[FoundCycloneResult]:
+    if "CYCLONEDDS_HOME" in os.environ:
+        dir = good_directory(Path(os.environ["CYCLONEDDS_HOME"]))
+        if dir:
+            return dir
+    if "CycloneDDS_ROOT" in os.environ:
+        dir = good_directory(Path(os.environ["CMAKE_CycloneDDS_ROOT"]))
+        if dir:
+            return dir
+    if "CycloneDDS_ROOT" in os.environ:
+        dir = good_directory(Path(os.environ["CycloneDDS_ROOT"]))
+        if dir:
+            return dir
+    if "CMAKE_PREFIX_PATH" in os.environ:
+        dir = search_cyclone_pathlike(os.environ["CMAKE_PREFIX_PATH"])
+        if dir:
+            return dir
+    if "CMAKE_LIBRARY_PATH" in os.environ:
+        dir = search_cyclone_pathlike(os.environ["CMAKE_LIBRARY_PATH"])
+        if dir:
+            return dir
+    if platform.system() != "Windows" and "LIBRARY_PATH" in os.environ:
+        dir = search_cyclone_pathlike(os.environ["LIBRARY_PATH"], upone=True)
+        if dir:
+            return dir
+    if platform.system() != "Windows" and "LD_LIBRARY_PATH" in os.environ:
+        dir = search_cyclone_pathlike(os.environ["LD_LIBRARY_PATH"], upone=True)
+        if dir:
+            return dir
+    if platform.system() == "Windows" and "PATH" in os.environ:
+        dir = search_cyclone_pathlike(os.environ["PATH"], upone=True)
+        if dir:
+            return dir
+

--- a/cyclonedds/__idlc__.py
+++ b/cyclonedds/__idlc__.py
@@ -10,21 +10,13 @@
  * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
 """
 
-from os import path, listdir
 from sys import exit
+from pathlib import Path
 
 
-dir = path.abspath(path.join(path.dirname(__file__), ".libs"))
-libs = [f for f in listdir(dir) if "idl" in f]
-
-if not libs:
-    idlpy_lib = None
-else:
-    idlpy_lib = path.join(dir, libs[0])
+idlpy_path = list(Path(__file__).resolve().parent.glob("_idlpy*"))[0]
 
 
 if __name__ == "__main__":
-    if not idlpy_lib:
-        exit(1)
-    print(idlpy_lib, end="", flush=True)
+    print(idlpy_path, end="", flush=True)
     exit(0)

--- a/cyclonedds/internal.py
+++ b/cyclonedds/internal.py
@@ -18,6 +18,7 @@ import ctypes as ct
 from ctypes.util import find_library
 from functools import wraps
 from dataclasses import dataclass
+from .__library__ import library_path, in_wheel
 
 
 class CycloneDDSLoaderException(Exception):
@@ -26,24 +27,19 @@ class CycloneDDSLoaderException(Exception):
 
 def _load(path):
     """Most basic loader, return the loaded DLL on path"""
-    library = ct.CDLL(path)
+    try:
+        library = ct.CDLL(path)
+    except OSError:
+        raise CycloneDDSLoaderException(f"Failed to load CycloneDDS library from {path}")
     if not library:
-        raise CycloneDDSLoaderException("Failed to load CycloneDDS library from {path}")
+        raise CycloneDDSLoaderException(f"Failed to load CycloneDDS library from {path}")
     return library
 
 
 def _loader_wheel_gen(rel_path, ext):
-    """
-        The tools 'auditwheel', 'delvewheel' and 'delocate' include libddsc into the Python environment in
-        a predictable location. This location differs slightly per tool. Generate a loader given the relative
-        path and platform-dependend extension.
-    """
     def _loader_wheel():
-        origin = os.path.join(os.path.dirname(__file__), *rel_path)
-        if os.path.exists(origin):
-            for file in os.listdir(origin):
-                if "ddsc" in file and ext in file and file.index(ext) > file.index("ddsc"):
-                    return _load(os.path.join(origin, file))
+        if in_wheel:
+            return _load(str(library_path))
         return None
     return _loader_wheel
 
@@ -82,21 +78,32 @@ def _loader_on_path_gen(name):
     return _loader_on_path
 
 
+
+def _loader_install_path():
+        try:
+            return _load(str(library_path))
+        except CycloneDDSLoaderException:
+            pass
+        return None
+
 _loaders_per_system = {
     "Linux": [
         _loader_wheel_gen(["..", "cyclonedds.libs"], ".so"),
-        _loader_cyclonedds_home_gen("lib/libddsc.so"),
+        _loader_cyclonedds_home_gen(f"lib{os.sep}libddsc.so"),
         _loader_on_path_gen("libddsc.so"),
+        _loader_install_path
     ],
     "Windows": [
         _loader_wheel_gen(["..", "cyclonedds.libs"], ".dll"),
-        _loader_cyclonedds_home_gen("bin\\ddsc.dll"),
-        _loader_on_path_gen("ddsc.dll")
+        _loader_cyclonedds_home_gen(f"bin{os.sep}ddsc.dll"),
+        _loader_on_path_gen("ddsc.dll"),
+        _loader_install_path
     ],
     "Darwin": [
         _loader_wheel_gen([".dylibs"], ".dylib"),
-        _loader_cyclonedds_home_gen("lib/libddsc.dylib"),
+        _loader_cyclonedds_home_gen(f"lib{os.sep}libddsc.dylib"),
         _loader_on_path_gen("libddsc.dylib"),
+        _loader_install_path
     ]
 }
 
@@ -124,7 +131,6 @@ def load_cyclonedds() -> ct.CDLL:
         "The CycloneDDS library could not be located. "
         "Try setting the CYCLONEDDS_HOME variable to what you used as CMAKE_INSTALL_PREFIX."
     )
-
 
 def c_call(cname):
     """

--- a/cyclonedds/tools/wheel_idlc/__init__.py
+++ b/cyclonedds/tools/wheel_idlc/__init__.py
@@ -11,36 +11,33 @@
 
 
 This tool is used in wheels to perform the idlc command.
-The idlc binary is under cyclonedds/.bin/idlc, libs are under
+The idlc binary is under cyclonedds/.libs/idlc, libs are under
 cyclonedds/.libs.
 """
 
 import os
 import sys
 import platform
-import subprocess
 import cyclonedds
+from pathlib import Path
 
-basedir = os.path.abspath(os.path.dirname(cyclonedds.__file__))
-idlc = os.path.join(basedir, ".bin", "idlc")
-if platform.system() == "Windows":
-    idlc += ".exe"
 
-libdir = os.path.join(basedir, ".libs")
+libdir = Path(__file__).resolve().parent / '.libs'
+idlc = (libdir / 'idlc.exe') if platform.system() == "Windows" else (libdir / 'idlc')
 
 
 def command():
-    if not os.path.exists(idlc):
+    if not idlc.exists():
         print("Python idlc entrypoint active but cyclonedds-python installation does not include idlc executable!")
         sys.exit(1)
 
     environ = os.environ.copy()
 
     if platform.system() == "Windows":
-        environ["PATH"] = ";".join([libdir] + environ.get("PATH", "").split(";"))
+        environ["PATH"] = ";".join([str(libdir)] + environ.get("PATH", "").split(";"))
     elif platform.system() == "Darwin":
-        environ["DYLD_LIBRARY_PATH"] = ":".join([libdir] + environ.get("DYLD_LIBRARY_PATH", "").split(":"))
+        environ["DYLD_LIBRARY_PATH"] = ":".join([str(libdir)] + environ.get("DYLD_LIBRARY_PATH", "").split(":"))
     else:
-        environ["LD_LIBRARY_PATH"] = ":".join([libdir] + environ.get("LD_LIBRARY_PATH", "").split(":"))
+        environ["LD_LIBRARY_PATH"] = ":".join([str(libdir)] + environ.get("LD_LIBRARY_PATH", "").split(":"))
 
     os.execvpe(idlc, sys.argv[1:], environ)

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -3,11 +3,8 @@ requires-python = ">=3.7"
 
 [build-system]
 requires = [
-    "setuptools>=42",
-    "wheel",
-    "scikit-build",
-    "cmake>=3.18.0",
-    "ninja"
+    "setuptools>=42.0.2",
+    "wheel>=0.29.0"
 ]
 build-backend = "setuptools.build_meta"
 

--- a/setup.py
+++ b/setup.py
@@ -1,6 +1,7 @@
 #!/usr/bin/env python
+
 """
- * Copyright(c) 2021 ADLINK Technology Limited and others
+ * Copyright(c) 2022 ZettaScale Technology and others
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0 which is available at
@@ -11,46 +12,80 @@
  * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
 """
 
-import re
-from os import environ
+import os
+import sys
 from pathlib import Path
-from skbuild import setup
-from setuptools import find_packages
-
+from setuptools import setup, Extension, find_packages
 
 this_directory = Path(__file__).resolve().parent
+sys.path.insert(0, str(this_directory / 'buildhelp'))
+
+from cyclone_search import find_cyclonedds
+from build_ext import build_ext, Library
+from bdist_wheel import bdist_wheel
+
+
 with open(this_directory / 'README.md', encoding='utf-8') as f:
     long_description = f.read()
 
 
-# invalidate cmake cache
-for cache_file in (this_directory / "_skbuild").rglob("CMakeCache.txt"):
-    cache_file.write_text(
-        re.sub("^//.*$\n^[^#].*pip-build-env.*$", "", cache_file.read_text(), flags=re.M)
+cyclone = find_cyclonedds()
+
+if not cyclone:
+    print("Could not locate cyclonedds. Try to set CYCLONEDDS_HOME or CMAKE_PREFIX_PATH")
+    import sys
+    sys.exit(1)
+
+
+with open(this_directory / 'cyclonedds' / '__library__.py', "w", encoding='utf-8') as f:
+    f.write("in_wheel = False\n")
+    f.write(f"library_path = '{cyclone.ddsc_library}'")
+
+
+ext_modules = [
+    Extension('cyclonedds._clayer', [
+            'clayer/cdrkeyvm.c',
+            'clayer/pysertype.c',
+            'clayer/typeser.c'
+        ],
+        include_dirs=[
+            str(cyclone.include_path),
+            str(this_directory / "clayer")
+        ],
+        libraries=['ddsc'],
+        library_dirs=[
+            str(cyclone.library_path),
+            str(cyclone.binary_path),
+        ]
     )
-
-console_scripts = [
-    "ddsls=cyclonedds.tools.ddsls:command",
-    "pubsub=cyclonedds.tools.pubsub:command"
 ]
-cmake_args = []
 
-if "CIBUILDWHEEL" in environ:
-    # We are building wheels! This means we should be including the idl compiler in the
-    # resulting package. To do this we need to include the idlc executable and libidl,
-    # this is done by cmake. We will add an idlc entrypoint that will make sure the load paths
-    # of idlc are correct.
-    console_scripts.append("idlc=cyclonedds.tools.wheel_idlc:command")
-    cmake_args.append("-DCIBUILDWHEEL=1")
-
-
-if "CYCLONEDDS_HOME" in environ and "CMAKE_PREFIX_PATH" not in environ:
-    cmake_args.append(f"-DCMAKE_PREFIX_PATH=\"{environ['CYCLONEDDS_HOME']}\"")
+if cyclone.idlc_library:
+    ext_modules += [
+        Library('cyclonedds._idlpy', [
+                'idlpy/src/context.c',
+                'idlpy/src/generator.c',
+                'idlpy/src/naming.c',
+                'idlpy/src/ssos.c',
+                'idlpy/src/types.c',
+                'idlpy/src/util.c'
+            ],
+            include_dirs=[
+                str(cyclone.include_path),
+                str(this_directory / "idlpy" / "include")
+            ],
+            libraries=['ddsc', 'cycloneddsidl'],
+            library_dirs=[
+                str(cyclone.library_path),
+                str(cyclone.binary_path)
+            ]
+        )
+    ]
 
 
 setup(
     name='cyclonedds',
-    version='0.8.0',
+    version='0.9.0',
     description='Eclipse Cyclone DDS Python binding',
     long_description=long_description,
     long_description_content_type="text/markdown",
@@ -79,13 +114,21 @@ setup(
         "Programming Language :: Python :: 3.10",
         "Operating System :: OS Independent"
     ],
-    packages=find_packages(".", exclude=("tests", "tests.*", "docs.*")),
+    packages=find_packages(".", include=("cyclonedds", "cyclonedds.*")),
     package_data={
         "cyclonedds": ["*.so", "*.dylib", "*.dll", "idlc*", "*py.typed"],
         "cyclonedds.idl": ["py.typed"]
     },
+    ext_modules=ext_modules,
+    cmdclass={
+        'bdist_wheel': bdist_wheel,
+        'build_ext': build_ext,
+    },
     entry_points={
-        "console_scripts": console_scripts,
+        "console_scripts": [
+            "ddsls=cyclonedds.tools.ddsls:command",
+            "pubsub=cyclonedds.tools.pubsub:command"
+        ],
     },
     python_requires='>=3.7',
     install_requires=[
@@ -106,6 +149,5 @@ setup(
             "sphinx-rtd-theme>=0.5.2"
         ]
     },
-    zip_safe=False,
-    cmake_args=cmake_args
+    zip_safe=False
 )

--- a/tests/test_loader.py
+++ b/tests/test_loader.py
@@ -1,5 +1,7 @@
 from cyclonedds.internal import CycloneDDSLoaderException, load_cyclonedds
+from cyclonedds.__library__ import library_path
 from pytest_mock import MockerFixture
+import os
 
 
 def gen_test_loader(loadlist):
@@ -20,10 +22,6 @@ def common_mocks(mocker: MockerFixture, platform: str, ext: str):
     mocker.patch("ctypes.util.find_library", new=lambda x: None)
     mocker.patch("cyclonedds.internal._load", new=gen_test_loader(loadlist))
     mocker.patch("platform.system", new=lambda: platform)
-    mocker.patch("os.path.join", new=gen_joiner("\\" if platform == "Windows" else "/"))
-    mocker.patch("os.path.exists", new=lambda p: True)
-    mocker.patch("os.listdir", new=lambda p: [f"libddsc_listdir_canary{ext}"])
-    mocker.patch("os.path.dirname", new=lambda f: "dirname_canary")
     mocker.patch("os.environ", new={"CYCLONEDDS_HOME": "env_canary", "PATH": ""})
     return loadlist
 
@@ -36,9 +34,9 @@ def test_loading_linux(mocker: MockerFixture):
         pass
 
     assert paths == [
-        "dirname_canary/../cyclonedds.libs/libddsc_listdir_canary.so",
-        "env_canary/lib/libddsc.so",
-        "libddsc.so"
+        f"env_canary{os.sep}lib{os.sep}libddsc.so",
+        "libddsc.so",
+        library_path
     ]
 
 
@@ -50,9 +48,9 @@ def test_loading_macos(mocker):
         pass
 
     assert paths == [
-        "dirname_canary/.dylibs/libddsc_listdir_canary.dylib",
-        "env_canary/lib/libddsc.dylib",
+        f"env_canary{os.sep}lib{os.sep}libddsc.dylib",
         "libddsc.dylib",
+        library_path
     ]
 
 
@@ -64,7 +62,7 @@ def test_loading_windows(mocker):
         pass
 
     assert paths == [
-        "dirname_canary\\..\\cyclonedds.libs\\libddsc_listdir_canary.dll",
-        "env_canary\\bin\\ddsc.dll",
-        "ddsc.dll"
+        f"env_canary{os.sep}bin{os.sep}ddsc.dll",
+        "ddsc.dll",
+        library_path
     ]

